velox/functions/sparksql/aggregates/AverageAggregate.cpp

```diff
@@ -0,0 +1,263 @@
+                             .build());
+  }
+  // Real input type in Presto has special case and returns REAL, not DOUBLE.
+  signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                           .returnType("real")
+                           .intermediateType("row(double,bigint)")
+                           .argumentType("real")
+                           .build());
+
+  signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                           .integerVariable("a_precision")
+                           .integerVariable("a_scale")
+                           .argumentType("DECIMAL(a_precision, a_scale)")
+                           .intermediateType("varbinary")
+                           .returnType("DECIMAL(a_precision, a_scale)")
+                           .build());
+
+  return exec::registerAggregateFunction(
+      name,
+      std::move(signatures),
+      [name](
+          core::AggregationNode::Step step,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& resultType,
+          const core::QueryConfig& /*config*/)
+          -> std::unique_ptr<exec::Aggregate> {
+        VELOX_CHECK_LE(
+            argTypes.size(), 1, "{} takes at most one argument", name);
+        auto inputType = argTypes[0];
+        if (exec::isRawInput(step)) {
+          switch (inputType->kind()) {
+            case TypeKind::SMALLINT:
+              return std::make_unique<
+                  AverageAggregate<int16_t, double, double>>(resultType);
+            case TypeKind::INTEGER:
+              return std::make_unique<
+                  AverageAggregate<int32_t, double, double>>(resultType);
+            case TypeKind::BIGINT: {
+              if (inputType->isShortDecimal()) {
+                return std::make_unique<DecimalAverageAggregateBase<int64_t>>(
+                    resultType);
+              }
+              return std::make_unique<
+                  AverageAggregate<int64_t, double, double>>(resultType);
+            }
+            case TypeKind::HUGEINT: {
+              if (inputType->isLongDecimal()) {
+                return std::make_unique<DecimalAverageAggregateBase<int128_t>>(
+                    resultType);
+              }
+              VELOX_NYI();
+            }
+            case TypeKind::REAL:
+              return std::make_unique<AverageAggregate<float, double, float>>(
+                  resultType);
+            case TypeKind::DOUBLE:
+              return std::make_unique<AverageAggregate<double, double, double>>(
+                  resultType);
+            default:
+              VELOX_FAIL(
+                  "Unknown input type for {} aggregation {}",
+                  name,
+                  inputType->kindName());
+          }
+        } else {
+          checkSumCountRowType(
+              inputType,
+              "Input type for final aggregation must be (sum:double/long decimal, count:bigint) struct");
```

Comment:
Is this any different from Presto?

Reply:
For Count type, Spark uses LongType which is the same with Presto. For Sum type, this is the same except for `YearMonthIntervalType` and `DayTimeIntervalType`, but these two are not supported yet.

```
  private lazy val sumDataType = child.dataType match {
    case _ @ DecimalType.Fixed(p, s) => DecimalType.bounded(p + 10, s)
    case _: YearMonthIntervalType => YearMonthIntervalType()
    case _: DayTimeIntervalType => DayTimeIntervalType()
    case _ => DoubleType
  }
```